### PR TITLE
feat: Observer (Fabric)

### LINKED ISSUE #24: Observer
Observer will listen to block events on its local peer on all configured channels. On each block event (transaction commit) on the channel it will check for Sidetree transaction.
Each Sidetree transaction will be processed as follows:
1) Resolve transaction by retrieving anchor and batch files from DCAS
2)  Store each operation into DID Document CC.

### PR DESCRIPTION
**Pull request by sandrask**
Observer will listen to block events on its local peer on all configured channels. On each block event (transaction commit) on the channel it will check for Sidetree transaction.
Each Sidetree transaction will be processed as follows:

- Resolve transaction by retrieving anchor and batch files from DCAS
- Store each operation into DID Document CC.

Closes #24

Signed-off-by: Sandra Vrtikapa <sandra.vrtikapa@securekey.com>

### CODE REVIEW

**Comment by sandrask** [comment]
verify-bddtest

**Comment by sandrask** [comment]
verify-bddtest

File: pkg/observer/observer.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+		}
+
+		logger.Debugf("successfully stored operation[%s] at address[%s]", op, addr)
+	}
+
+	return nil
+}
+
+var getDCAS = func(channelID string) dcasClient {
+	return dcasclient.New(channelID)
```
Comment (bstasyszyn) [inline pkg/observer/observer.go]: It's expensive to create a new client every time. The client maintains a cache of collection configs. The first time it's used, it needs to call the ledger to retrieve the collection config. It's better to reuse the client.

File: pkg/observer/observer.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+		}
+
+	} else {
+		logger.Info("peer is not an observer, nothing to do...")
+	}
+
+	return nil
+}
+
+func initObservingOnChannel(channelID string) {
```
Comment (bstasyszyn) [inline pkg/observer/observer.go]: It feels like there should be an Observer struct to encapsulate all of the observer functions, variables.

File: pkg/observer/observer.go (hunk 1 above)
Comment (sandrask) [reply pkg/observer/observer.go]: Done

File: pkg/observer/observer.go [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+			logger.Debugf("Successfully processed anchor[%s] on channel[%s]", txn.AnchorAddress, channel)
+
+		}
+	}(o.channelID, sidetreeTxnChannel)
+
+}
+
+func (o *observer) getDCASClient() dcasClient {
+
+	if o.dcas != nil {
```
Comment (bstasyszyn) [inline pkg/observer/observer.go]: You need to have an RLock() before reading o.dcas.

File: pkg/observer/observer.go [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+	dcas := getDCAS(o.channelID)
+
+	o.Lock()
+	defer o.Unlock()
+
+	o.dcas = dcas
+	return dcas
+}
+
+func (o *observer) processSidetreeTxn(channel string, sidetreeTxn notifier.SidetreeTxn) error {
```
Comment (bstasyszyn) [inline pkg/observer/observer.go]: You don't need a channel arg since you can get it from observer.

File: pkg/observer/observer.go [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+	af, err := getAnchorFile(content)
+	if err != nil {
+		return errors.Wrapf(err, "failed to unmarshal anchor[%s]", sidetreeTxn.AnchorAddress)
+	}
+
+	return o.processBatchFile(channel, af.BatchFileHash)
+}
+
+func (o *observer) processBatchFile(channel string, batchFileAddress string) error {
```
Comment (bstasyszyn) [inline pkg/observer/observer.go]: No need for channel arg since you can get it from observer

File: pkg/observer/observer.go (hunk 3 above)
Comment (sandrask) [reply pkg/observer/observer.go]: Added RLock()

File: pkg/observer/observer.go (hunk 5 above)
Comment (sandrask) [reply pkg/observer/observer.go]: Removed channel from args

File: pkg/observer/observer.go (hunk 4 above)
Comment (sandrask) [reply pkg/observer/observer.go]: Removed channel from args

File: pkg/observer/notifier/notifier.go [hunk 6]
```diff
@@ -38,24 +38,24 @@ func New(publisher blockPublisher) *Notifier {
...
 	BlockNumber   uint64
 	TxNum         uint64
 	AnchorAddress string
 }
 
 // RegisterForSidetreeTxn register to get AnchorFileAddress value from writeset in the block committed by sidetreetxn_cc
-func (n *Notifier) RegisterForSidetreeTxn() <-chan SideTreeTxn {
-	anchorFileAddressChan := make(chan SideTreeTxn, 100)
+func (n *Notifier) RegisterForSidetreeTxn() <-chan SidetreeTxn {
+	anchorFileAddressChan := make(chan SidetreeTxn, 100)
```
Comment (troyronda) [inline pkg/observer/notifier/notifier.go]: why 100? (I see this was from a previous commit but still curious)

File: pkg/observer/notifier/notifier_test.go [hunk 7]
```diff
@@ -47,14 +47,14 @@ func TestRegisterForAnchorFileAddress(t *testing.T) {
 	t.Run("test namespace not equal to sideTreeTxnCCName", func(t *testing.T) {
 		// register to receive sidetree txn value
 		sideTreeTxnCh := notifier.RegisterForSidetreeTxn()
-		done := make(chan SideTreeTxn, 1)
+		done := make(chan SidetreeTxn, 1)
```
Comment (troyronda) [inline pkg/observer/notifier/notifier_test.go]: note: a done channel can be simply:

make(chan struct{})

**Pull request merged**